### PR TITLE
Feature: Horizontal lists

### PR DESCRIPTION
**Pull request by frysch**
Adding possibility to create horizontal drag n' drop areas.

### CODE REVIEW

File: addon/components/drag-sort-item.js [hunk 1]
```diff
@@ -65,10 +68,13 @@ export default Component.extend({
   isDragged2     : false,
   originalHeight : null,
 
-  shouldShowPlaceholderAbove2 : undefined,
-  shouldShowPlaceholderBelow2 : undefined,
-
-
+  shouldShowPlaceholderAbove2     : undefined,
+  shouldShowPlaceholderBelow2     : undefined,
+  shouldShowPlaceholderAbove3     : and('shouldShowPlaceholderAbove2', 'notHorizontal'),
```
[the remark this answers is not in the document]
Comment (frysch) [reply addon/components/drag-sort-item.js]: I can't see how it should work with that solution. Of course it's possible if i just remove the CSS classes `-placeholderLeft` and `-placeholderRight` and just style based on the `-horizontal` class on `.dragSortList` instead. The `shouldShowPlaceholderAbove` needs to return true when the placeholder should display left. So maybe renaming the property to something like `shouldShowPlaceholderBefore` would make it clearer?

**Comment by frysch** [comment]
Hey!

1. Maybe the bug's still there, dunno. If it is, please tell me which browser you're using.

2. I've changed it to use flexbox instead.

3. No scrollbars, it'll have to be customized if and when someone needs it.

4. I had the same thoughts, can't see what's more to test :)

File: addon/components/drag-sort-item.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (frysch) [reply addon/components/drag-sort-item.js]: Aaah, sorry. Github hid some more comments. You're thinking just like me further down the comments :)

**Comment by frysch** [comment]
Ping @lolmaus  👆(Sorry if i'm spamming you with emails now. Dunno what notification settings you have)

File: addon/components/drag-sort-item.js [hunk 2]
```diff
@@ -219,21 +220,34 @@ export default Component.extend({
...
-    const isPlaceholderBelow = this.get('shouldShowPlaceholderBelow2')
+    const group        = this.get('group')
+    const index        = this.get('index')
+    const items        = this.get('items')
+    const element      = this.get('element')
+    const isHorizontal = this.get('dragSort.isHorizontal')
+
+    const isPlaceholderBefore   = isHorizontal
+      ? this.get('shouldShowPlaceholderToTheLeft')
+      : this.get('shouldShowPlaceholderBefore2')
```
Comment (lolmaus) [inline addon/components/drag-sort-item.js]: A very minor request: please format ternaries like this:

```js
    const isPlaceholderBefore =
      isHorizontal
        ? this.get('shouldShowPlaceholderToTheLeft')
        : this.get('shouldShowPlaceholderBefore2')
```

File: addon/components/drag-sort-item.js [hunk 3]
```diff
@@ -219,21 +226,33 @@ export default Component.extend({
...
+
+    const isPlaceholderAfter    = isHorizontal
+      ? this.get('shouldShowPlaceholderToTheRight')
+      : this.get('shouldShowPlaceholderBelow2')
 
     const placeholderCorrection =
-      isPlaceholderAbove ?  getComputedStyleInt(element, 'padding-top')    :
-      isPlaceholderBelow ? -getComputedStyleInt(element, 'padding-bottom') :
+      isPlaceholderBefore ?  getComputedStyleInt(element, isHorizontal ? 'padding-left' : 'padding-top')  :
+      isPlaceholderAfter ? -getComputedStyleInt(element, isHorizontal ? 'padding-right' : 'padding-bottom') :
```
[the remark this answers is not in the document]
Comment (lolmaus) [reply addon/components/drag-sort-item.js]: This is still misaligned.

My take on alignment:

```js
    const placeholderCorrection =
      isPlaceholderBefore ?  getComputedStyleInt(element, isHorizontal ? 'padding-left'  : 'padding-top')  :
      isPlaceholderAfter  ? -getComputedStyleInt(element, isHorizontal ? 'padding-right' : 'padding-bottom') :
                             0  // eslint-disable-line indent
```

I also don't like the way ternaries are nested. How about this?

```js
    const placeholderCorrection =
      ( isPlaceholderBefore &&  isHorizontal) ? getComputedStyleInt(element, 'padding-left')   :
      ( isPlaceholderBefore && !isHorizontal) ? getComputedStyleInt(element, 'padding-top')    :
      (!isPlaceholderBefore &&  isHorizontal) ? getComputedStyleInt(element, 'padding-right')  :
      (!isPlaceholderBefore && !isHorizontal) ? getComputedStyleInt(element, 'padding-bottom') :
                                                0 // eslint-disable-line indent
```

File: addon/components/drag-sort-item.js [hunk 4]
```diff
@@ -47,8 +48,8 @@ export default Component.extend({
     ':dragSortItem',
     'isDragged2:-isDragged',
     'isDraggingOver:-isDraggingOver',
-    'shouldShowPlaceholderAbove2:-placeholderAbove',
-    'shouldShowPlaceholderBelow2:-placeholderBelow',
+    'shouldShowPlaceholderBefore2:-placeholderAbove',
```
Comment (lolmaus) [inline addon/components/drag-sort-item.js]: You renamed the JS property, but the HTML class is still using the old name. Please update from above/below to before/after for consistency.

File: addon/styles/addon.css [hunk 5]
```diff
@@ -61,3 +70,27 @@
   /*top: 100%;*/
   bottom: 0;
 }
+
+.-horizontal .dragSortItem.-placeholderAbove:before,
+.-horizontal .dragSortItem.-placeholderBelow:before {
+  text-align: center;
+  width: 25px;
+  top: 0;
```
Comment (lolmaus) [inline addon/styles/addon.css]: `top: 0` fails to center the placeholder arrow vertically.

This should do vertical centering:

```css
.-horizontal .dragSortItem.-placeholderAbove:before,
.-horizontal .dragSortItem.-placeholderBelow:before {
  top: 50%;
  transform: translateY(-50%);
}
```

File: addon/styles/addon.css [hunk 6]
```diff
@@ -44,8 +43,18 @@
   padding-bottom: 25px;
 }
 
+.-horizontal .dragSortItem.-placeholderAbove {
```
Comment (lolmaus) [inline addon/styles/addon.css]: You should never be using modifier classes (that start with `-`) individually.

They must always come paired with a semantic class like this: `.dragSortList.-horizontal`.

File: addon/components/drag-sort-item.js (hunk 4 above)
Comment (frysch) [reply addon/components/drag-sort-item.js]: Actually choose to not rename it. Didn't want to break styles for those already using the component. But, if you're fine with that, i'm happy to change the classes.

Comment (lolmaus) [reply addon/components/drag-sort-item.js]: I don't mind a breaking change, but that adds a requirement to update the readme... 😇

Comment (frysch) [reply addon/components/drag-sort-item.js]: I'll change it then 👍

File: addon/components/drag-sort-item.js [hunk 7]
```diff
@@ -65,10 +67,9 @@ export default Component.extend({
...
-  shouldShowPlaceholderAbove2 : undefined,
-  shouldShowPlaceholderBelow2 : undefined,
-
-
+  shouldShowPlaceholderBefore2 : undefined,
+  shouldShowPlaceholderAfter2  : undefined,
+  isVertical                   : not('dragSort.isHorizontal'),
 
   // ----- Aliases -----
   isDraggingUp : reads('dragSort.isDraggingUp'),
```
Comment (lolmaus) [inline addon/components/drag-sort-item.js]: This name is now improper. But I can't think of a better name. Let's keep it as it is.

File: tests/dummy/app/controllers/index.js [hunk 8]
```diff
@@ -103,6 +103,24 @@ export default Controller.extend({
...
+      {name : 'Quux'},
+    ])
+  ),
+
+  items14 : computed(() =>
+    A([
+      {name : 'أخرى ماشاء أي مكن'},
+      {name : 'أضف ما معقل'},
+      {name : 'أخر أحدث والمانيا'},
+      {name : '٣٠ هذا أوسع أوروبا ليتسنّى'},
```
[the remark this answers is not in the document]
Comment (frysch) [reply tests/dummy/app/controllers/index.js]: I just google something like `lorem ipsum arabic` and found this. Have no idea what it is really, but hey, let's do the Lord of the Rings :)

**Comment by frysch** [comment]
> Hey man, the demo was good, I did not want you to remove it.
> 
> Just add a warning like "There may be issues with nested horizontal lists. Before making a decision to use it from this addon, please fiddle with this demo to see if it's okay for you".

Actually it was behaving very strange. Items disappearing and after dropping them, and items not being draggable at all. I think it needs some lovin' if it should be included.

**Comment by lolmaus** [comment]
OK, got it.

**Comment by frysch** [comment]
Hey!

I applied the suggested code in `addon/components/drag-sort-list.js`. Worked well, nice work! But i tweaked it a bit to target the last item on the hovered row instead of only target the last item. Think it turned out really nice! 😀

There might be an idea to actually do something similar on vertical lists as well i think. I got a case at work right now where we have three lists wrapped in a flex div, which makes em equal high. So when we're dragging onto a list the items suggested spot is not where one expect. But, that's more a suggestion for the future. Not for this PR.

File: addon/components/drag-sort-list.js [hunk 9]
```diff
@@ -147,22 +169,27 @@ export default Component.extend({
...
+    const itemsNodeList      = this.get('element').querySelectorAll('.dragSortItem')
+    const draggableItems     = A(Array.prototype.slice.call(itemsNodeList))
+    const positions          = A(draggableItems.map(draggableItem => draggableItem.getBoundingClientRect()))
+    const rows               = positions.uniqBy('top').mapBy('top').sort()
+    const currentRowPosition = rows.filter(row => row < event.clientY).pop()
+    const closestItem        = positions.filterBy('top', currentRowPosition).pop()
+
+    return closestItem
+      ? positions.indexOf(closestItem)
+      : 0
```
Comment (lolmaus) [inline addon/components/drag-sort-list.js]: This can be a one-liner.

File: addon/components/drag-sort-list.js [hunk 10]
```diff
@@ -123,21 +130,64 @@ export default Component.extend({
...
+    if (isHorizontal) {
+      targetIndex = this.getClosestHorizontalIndex(event)
+      dragSort.set('isDraggingUp', false)
+    }
 
-    dragSort.dragEntering({group, items})
+    dragSort.dragEntering({group, items, isHorizontal, targetIndex})
+  },
+
+  getClosestHorizontalIndex (event) {
```
Comment (lolmaus) [inline addon/components/drag-sort-list.js]: This technique works very well! Fantastic!!

**Comment by frysch** [comment]
> Please make these final touches, and I'm merging and releasing it. 🙇

**Comment by lolmaus** [comment]
minions flipping out

**Comment by lolmaus** [comment]
it is done

Congrats! I'll release later today.

**Comment by emilkarl** [comment]
Nice that this is merged, when will it be released?

**Comment by lolmaus** [comment]
Released as `2.0.0-beta.0`. Congrats and thanks to @frysch!

**Pull request merged**